### PR TITLE
CI, BLD, TST: Re-enable Emscripten/Pyodide CI job for NumPy

### PR DESCRIPTION
**Pull request by agriyakhetarpal**
## Description

This PR supersedes gh-24603. It adds a CI job to test NumPy v2.0.0.dev0 against a Pyodide (wasm32) runtime. Some of the key changes here are:

1. A patch has been added to the `ci/tools/emscripten/` directory based on an upstream change to Pyodide at https://github.com/pyodide/pyodide/pull/4502. This patch ensures that the correct Meson build system (i.e., `vendored-meson`) is found during the build process.
2. A new requirements file has been created that contains pure-Python dependencies that can be installed inside a Pyodide virtual environment. It has been placed in the same directory as above.
3. Some ancillary files have been added to ensure that the WASM wheel is compiled without specialised CPU instructions (SIMD, etc.) and that the build procedure does not need to build against BLAS or LAPACK, which are currently unavailable on WebAssembly.
4. Various tests have been skipped, such as:
    - All `f2py` tests, since Fortran cannot run in WASM
    - Some tests for `np.where()` that require floating point exception support
    - Some tests related to the use of subprocesses to retrieve configurations for how NumPy has been built
    - A bug in Cython, i.e., https://github.com/cython/cython/issues/5411, has now been resolved. The `import cython` line was breaking the test discovery because Cython is not supported in Pyodide in-tree yet.
    - and so on

Thanks to @rgommers and the notes provided on gh-24603, all tests pass! Here's a workflow run from my fork where they can be observed: https://github.com/agriyakhetarpal/numpy/actions/runs/8065629246

### CODE REVIEW

**Comment by agriyakhetarpal** [comment]
Pyodide build is passing here as well: https://github.com/numpy/numpy/actions/runs/8066720365

**Comment by agriyakhetarpal** [comment]
It looks like the build with nightly OpenBLAS wheels is failing – is there something I can do to help here? (cc: @mattip)

**Comment by agriyakhetarpal** [comment]
> RE:Blas Is this based on the current tip of NumPy main?

Yes, this has been rebased on top of the `main` branch, IIUC and contains the latest changes.

**Comment by ngoldbaum** [comment]
The BLAS failures are happening on every PR and are unrelated.

File: meson_cpu/meson.build [hunk 1]
```diff
@@ -92,30 +92,27 @@ min_features = {
...
 }.get(cpu_family, {})
 max_features = []
 foreach fet_name, fet_obj : max_features_dict
   max_features += [fet_obj]
 endforeach
-if max_features.length() == 0
-  message('Disabling CPU feature detection due to unsupported architecture: "' + cpu_family + '"')
-  CPU_CONF_BASELINE = 'none'
-  CPU_CONF_DISPATCH = 'none'
-endif
```
Comment (ngoldbaum) [inline meson_cpu/meson.build:118]: Can you explain a little while you removed this default case?

**Review by ngoldbaum — COMMENTED** [review]
All the changes to tests make sense. I have a slight preference for marking tests with `xfail(strict=True)` for cases where we expect the test might pass in the future. I didn’t review the github actions changes but the job passes so probably close enough :)

File: .github/workflows/emscripten.yml [hunk 2]
```diff
@@ -18,51 +18,67 @@ permissions:
...
-        run: pip install "pydantic<2" pyodide-build==$PYODIDE_VERSION
+        run: pip install "pydantic<2" pyodide-build==${{ env.PYODIDE_VERSION }}
+
+      - name: Find installation for pyodide-build
+        shell: python
+        run: |
+          import os
+          import pyodide_build
+
+          pyodide_build_path = pyodide_build.__file__[:-12]
```
Comment (hoodmane) [inline .github/workflows/emscripten.yml:65]:
```suggestion
          import os
          import pyodide_build
          from pathlib import Path

          pyodide_build_path = str(Path(pyodide_build.__file__).parent)
```

File: meson_cpu/meson.build (hunk 1 above)
Comment (agriyakhetarpal) [reply meson_cpu/meson.build:118]: Oops, I can't seem to recall why I removed the default case. I'm happy to add this back again if this is needed – I'm not sure I understand what this code snippet does though

File: .github/workflows/emscripten.yml (hunk 2 above)
Comment (agriyakhetarpal) [reply .github/workflows/emscripten.yml:65]: Ah, good old `pathlib`, thanks! Much neater

**Comment by agriyakhetarpal** [comment]
> All the changes to tests make sense. I have a slight preference for marking tests with `xfail(strict=True)` for cases where we expect the test might pass in the future.

I just checked; I have been mostly conforming to the `pytest` skip markers that were already present. Out of the tests, Fortran (`f2py`) support should be unlikely to come up anytime soon and can be skipped entirely, but the floating-point exception support could be marked with `@pytest.mark.xfail(strict=True)`. However, I am not aware of aware of the developments or the effort required to make them possible on the Pyodide side of things.

**Review by ryanking13 — COMMENTED** [review]
Looks good to me for pyodide-build side. Thanks for working on this @agriyakhetarpal!

File: .github/workflows/emscripten.yml [hunk 3]
```diff
@@ -18,51 +18,68 @@ permissions:
...
-          # A shallow clone can work when the following issue is resolved:
-          # https://github.com/actions/checkout/issues/338
-          fetch-depth: 0
-
-      - name: set up python
+          submodules: recursive
+          # The fetch-tags input shall fetch tags for versioneer.py
+          # without the need to fetch the entire history, see
+          # https://github.com/actions/checkout#usage
+          fetch-tags: true
```
Comment (rgommers) [inline .github/workflows/emscripten.yml:42]: Nice change! This seems to save about 15 seconds of job runtime and a whole bunch of data transfer. We have lots of other jobs using `fetch-depth: 0`. Do you want to open a separate PR to make this change in all jobs?

File: .github/workflows/emscripten.yml [hunk 4]
```diff
@@ -18,51 +18,68 @@ permissions:
...
-          submodules: true
-          # versioneer.py requires the latest tag to be reachable. Here we
-          # fetch the complete history to get access to the tags.
-          # A shallow clone can work when the following issue is resolved:
-          # https://github.com/actions/checkout/issues/338
-          fetch-depth: 0
-
-      - name: set up python
+          submodules: recursive
+          # The fetch-tags input shall fetch tags for versioneer.py
```
Comment (rgommers) [inline .github/workflows/emscripten.yml:39]: Fetching tags is still needed, but we removed `versioneer` a while back. It's now a simply script that just invokes `git`, see lines 4-6 of the main `meson.build` file. Could you trim the comment a bit and remove the mention of versioneer?

File: meson_cpu/meson.build [hunk 5]
```diff
@@ -92,30 +92,27 @@ min_features = {
...
-  'aarch64': [ASIMD]
+  'aarch64': [ASIMD],
+  'wasm32': [],
 }.get(cpu_family, [])
 if host_machine.endian() == 'little' and cpu_family == 'ppc64'
   min_features = [VSX2]
 endif
 
-# Used by build option 'max/native/detect'
+# Used by build option 'max'
```
Comment (rgommers) [inline meson_cpu/meson.build:102]: I don't think this needed changing either?

File: meson_cpu/meson.build (hunk 1 above)
Comment (rgommers) [reply meson_cpu/meson.build:118]: Yes, please revert these changes; it looks to me like they should not be touched, and if they are causing a problem then let's make sure to understand that.

The warning is saying that we don't have SIMD support for this CPU type (correct); the two variables are used to determine SIMD-related compiler flags.

**Review by rgommers — COMMENTED** [review]
This looks good, nice. A few more small comments/requests.

> the floating-point exception support could be marked with `@pytest.mark.xfail(strict=True)`

I'd prefer not to do that. There are already lots of other identical skips for both floating-point errors and subprocesses, so keeping these the same is both nicer and less churn.

File: .github/workflows/emscripten.yml (hunk 3 above)
Comment (agriyakhetarpal) [reply .github/workflows/emscripten.yml:42]: Sure!

File: .github/workflows/emscripten.yml (hunk 4 above)
Comment (agriyakhetarpal) [reply .github/workflows/emscripten.yml:39]: Thanks for the information, I shall remove the comment about `versioneer` in here itself, and in a follow-up, remove any references to it that may have been missed.

File: meson_cpu/meson.build (hunk 5 above)
Comment (agriyakhetarpal) [reply meson_cpu/meson.build:102]: Will revert – I am unsure where this crept in from

File: .github/workflows/emscripten.yml (hunk 3 above)
Comment (agriyakhetarpal) [reply .github/workflows/emscripten.yml:41]: Addressed in gh-25900

**Comment by agriyakhetarpal** [comment]
I have resolved all the review comments that were posted; this should now be polished enough for another review, or for merging as and when needed.

**Review by rgommers — APPROVED** [review]
All looks good now. I did a slight cleanup of the commits that reverted something. In it goes. Nice work @agriyakhetarpal, and thanks for the reviews @hoodmane, @ryanking13 and @ngoldbaum.
